File: restaurant/service/order_service.py
from model.order import Order
from model.order_item import OrderItem
from model.enums import OrderStatus
from repository.order_repo import OrderRepository

class OrderService:
    """Service quản lý các đơn hàng"""
    
    def __init__(self, order_repo: OrderRepository):
        self.order_repo = order_repo
        self._order_counter = 0
    
    def create_order(self, table_id: int) -> Order:
        """Tạo một đơn hàng mới"""
        self._order_counter += 1
        order = Order(self._order_counter, table_id)
        if self.order_repo.add_order(order):
            return order
        return None
    
    def get_order(self, order_id: int) -> Order:
        """Lấy thông tin đơn hàng"""
        return self.order_repo.get_order(order_id)
    
    def add_item_to_order(self, order_id: int, order_item: OrderItem) -> bool:
        """Thêm mục vào đơn hàng"""
        order = self.order_repo.get_order(order_id)
        if order:
            order.add_item(order_item)
            return self.order_repo.update_order(order)
        return False
    
    def remove_item_from_order(self, order_id: int, order_item_id: int) -> bool:
        """Xóa mục khỏi đơn hàng"""
        order = self.order_repo.get_order(order_id)
        if order:
            order.remove_item(order_item_id)
            return self.order_repo.update_order(order)
        return False
    
    def change_order_status(self, order_id: int, new_status: OrderStatus) -> bool:
        """Thay đổi trạng thái đơn hàng"""
        order = self.order_repo.get_order(order_id)
        if order:
            order.change_status(new_status)
            return self.order_repo.update_order(order)
        return False
    
    def get_all_orders(self) -> list[Order]:
        """Lấy tất cả đơn hàng"""
        return self.order_repo.get_all_orders()
    
    def get_orders_by_table(self, table_id: int) -> list[Order]:
        """Lấy các đơn hàng của một bàn"""
        return self.order_repo.get_orders_by_table(table_id)

File: backend/src/main/java/com/restaurant/service/OrderService.java
package com.restaurant.service;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.restaurant.model.entity.MenuItem;
import com.restaurant.model.entity.Order;
import com.restaurant.model.entity.OrderItem;
import com.restaurant.model.entity.Table;
import com.restaurant.repository.MenuRepository;
import com.restaurant.repository.OrderRepository;
import com.restaurant.repository.TableRepository;

@Service
@Transactional
public class OrderService {

    @Autowired
    private OrderRepository orderRepository;

    @Autowired
    private TableRepository tableRepository;

    @Autowired
    private MenuRepository menuRepository;

    public Order createOrderForTable(int tableId) {
        Table table = tableRepository.findById(tableId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy bàn với ID: " + tableId));
        Optional<Order> existingOrder = orderRepository.findActiveOrderByTableId(tableId);
        return existingOrder.orElseGet(() -> {
            Order newOrder = new Order(table);
            return orderRepository.save(newOrder);
        });
    }

    public Optional<Order> getActiveOrderByTable(int tableId) {
        return orderRepository.findActiveOrderByTableId(tableId);
    }

    public Order addItemToOrder(Long orderId, int menuItemId, int quantity) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));
        MenuItem menuItem = menuRepository.findById(menuItemId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy món với ID: " + menuItemId));
        
        OrderItem orderItem = new OrderItem(menuItem, quantity);
        order.addItem(orderItem);
        order.updateOrderStatus();
        return orderRepository.save(order);
    }

    // [CẬP NHẬT] Hàm này giờ sẽ xử lý cả Note
    public Order addMultipleItems(Long orderId, List<ItemRequest> itemRequests) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));

        itemRequests.forEach(request -> {
            MenuItem menuItem = menuRepository.findById(request.getMenuItemId())
                    .orElseThrow(() -> new RuntimeException("Không tìm thấy món với ID: " + request.getMenuItemId()));
            
            OrderItem orderItem = new OrderItem(menuItem, request.getQuantity());
            
            // [MỚI] Lưu ghi chú nếu có
            if (request.getNote() != null) {
                orderItem.setNote(request.getNote());
            }
            
            order.addItem(orderItem);
        });

        order.updateOrderStatus();
        return orderRepository.save(order);
    }

    public Order changeItemState(Long orderId, Long orderItemId) {
        Order order = orderRepository.findByIdWithItems(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));
        OrderItem orderItem = order.getOrderItems().stream()
                .filter(item -> item.getOrderItemId().equals(orderItemId))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("Không tìm thấy món với ID: " + orderItemId));
        orderItem.changeToNextState();
        order.updateOrderStatus();
        return orderRepository.save(order);
    }

    public Order changeAllItemsState(Long orderId) {
        Order order = orderRepository.findByIdWithItems(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));
        order.getOrderItems().forEach(OrderItem::changeToNextState);
        order.updateOrderStatus();
        return orderRepository.save(order);
    }

    public List<Order> getAllOrders() { return orderRepository.findAll(); }
    public Optional<Order> getOrderById(Long orderId) { return orderRepository.findByIdWithItems(orderId); }
    public List<Order> getOrdersByTable(int tableId) { return orderRepository.findByTableId(tableId); }
    public List<Order> getInProgressOrders() { return orderRepository.findAllInProgressOrders(); }

    public Order completeOrder(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));
        order.completeOrder();
        return orderRepository.save(order);
    }

    public Order removeItemFromOrder(Long orderId, Long orderItemId) {
        Order order = orderRepository.findByIdWithItems(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy order với ID: " + orderId));
        OrderItem itemToRemove = order.getOrderItems().stream()
                .filter(item -> item.getOrderItemId().equals(orderItemId))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("Không tìm thấy món với ID: " + orderItemId));
        order.removeItem(itemToRemove);
        order.updateOrderStatus();
        return orderRepository.save(order);
    }

    public void deleteOrder(Long orderId) {
        orderRepository.deleteById(orderId);
    }

    // [CẬP NHẬT] Inner Class thêm trường note
    public static class ItemRequest {
        private int menuItemId;
        private int quantity;
        private String note; // [MỚI]

        public ItemRequest() {}

        public ItemRequest(int menuItemId, int quantity, String note) {
            this.menuItemId = menuItemId;
            this.quantity = quantity;
            this.note = note;
        }

        public int getMenuItemId() { return menuItemId; }
        public void setMenuItemId(int menuItemId) { this.menuItemId = menuItemId; }

        public int getQuantity() { return quantity; }
        public void setQuantity(int quantity) { this.quantity = quantity; }
        
        public String getNote() { return note; }
        public void setNote(String note) { this.note = note; }
    }
}